reconcile/endpoints_discovery/integration.py

```diff
@@ -0,0 +1,347 @@
+from reconcile.utils.defer import defer
+from reconcile.utils.differ import diff_any_iterables
+from reconcile.utils.disabled_integrations import integration_is_enabled
+from reconcile.utils.extended_early_exit import (
+    ExtendedEarlyExitRunnerResult,
+    extended_early_exit_run,
+)
+from reconcile.utils.oc_map import OCMap, init_oc_map_from_namespaces
+from reconcile.utils.ruamel import create_ruamel_instance
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.unleash import get_feature_toggle_state
+from reconcile.utils.vcs import VCS
+
+QONTRACT_INTEGRATION = "endpoints-discovery"
+QONTRACT_INTEGRATION_VERSION = make_semver(1, 0, 0)
+
+
+class EndpointsDiscoveryIntegrationParams(PydanticRunParams):
+    thread_pool_size: int = 10
+    internal: bool | None = None
+    use_jump_host: bool = True
+    cluster_name: set[str] | None = None
+    namespace_name: str | None = None
+    endpoint_tmpl_resource: str = "/endpoints-discovery/endpoint-template.yml"
+    # extended early exit parameters
+    enable_extended_early_exit: bool = False
+    extended_early_exit_cache_ttl_seconds: int = 7200  # run every 2 hours
+    log_cached_log_output: bool = False
+
+
+class Route(BaseModel):
+    name: str
+    host: str
+    tls: bool
+
+    @property
+    def url(self) -> str:
+        return f"{self.host}:{443 if self.tls else 80}"
+
+
+def endpoint_prefix(namespace: NamespaceV1) -> str:
+    return f"{QONTRACT_INTEGRATION}-{namespace.cluster.name}-{namespace.name}"
+
+
+def compile_endpoint_name(endpoint_prefix: str, route: Route) -> str:
+    return f"{endpoint_prefix}-{route.name}"
+
+
+def render_template(template: str, endpoint_name: str, route: Route) -> dict:
+    yml = create_ruamel_instance()
+    return yml.load(
+        jinja2.Template(
+            template,
+            undefined=jinja2.StrictUndefined,
+            trim_blocks=True,
+            lstrip_blocks=True,
+            keep_trailing_newline=True,
+        ).render({"endpoint_name": endpoint_name, "route": route})
+    )
+
+
+class RunnerParams(TypedDict):
+    oc_map: OCMap
+    merge_request_manager: MergeRequestManager
+    endpoint_template: str
+    namespaces: Iterable[NamespaceV1]
+
+
+class EndpointsDiscoveryIntegration(
+    QontractReconcileIntegration[EndpointsDiscoveryIntegrationParams]
+):
+    """Discover routes from all OpenShift clusters and update endPoints in app-interface."""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_early_exit_desired_state(
+        self, query_func: Callable | None = None
+    ) -> dict[str, Any]:
+        """Return the desired state for early exit."""
+        if not query_func:
+            query_func = gql.get_api().query
+        return {
+            "namespaces": [ns.dict() for ns in self.get_namespaces(query_func)],
```

Comment:
there maybe some order issues on `namespaces`, better to use dict like `(ns.cluster.name, ns.name): ns.dict()` or sort them

Reply:
29a2620 removes it because the integration won't run during the PR checks.